Whenever I come across a new mathematical object, I try to see if it is implemented in [SageMath](https://www.sagemath.org/), an open-source free program that has implementations of various mathematical objects. It's a good playground to test various conjectures and get a feel for the objects that we are discussing in class. 

Here, since we have introduced what a Coxeter group is and since [Coxeter groups are implemented in Sage](https://doc.sagemath.org/html/en/reference/categories/sage/categories/coxeter_groups.html), we will play around with the irreducible finite Coxeter systems whose Coxeter diagrams are displayed in Fig 11.1. 

Note we first have the infinite families:

### $A_n$
Below, we construct the Coxeter group $A_4$:

In [1]:
A4 = CoxeterGroup(["A", 4] )
A4

Finite Coxeter group over Integer Ring with Coxeter matrix:
[1 3 2 2]
[3 1 3 2]
[2 3 1 3]
[2 2 3 1]

In [2]:
len(A4), list( A4 )

(120,
 [
[1 0 0 0]  [-1  1  0  0]  [ 1  0  0  0]  [ 1  0  0  0]  [ 1  0  0  0]
[0 1 0 0]  [ 0  1  0  0]  [ 1 -1  1  0]  [ 0  1  0  0]  [ 0  1  0  0]
[0 0 1 0]  [ 0  0  1  0]  [ 0  0  1  0]  [ 0  1 -1  1]  [ 0  0  1  0]
[0 0 0 1], [ 0  0  0  1], [ 0  0  0  1], [ 0  0  0  1], [ 0  0  1 -1],

[ 0 -1  1  0]  [-1  1  0  0]  [ 1  0  0  0]  [-1  1  0  0]
[ 1 -1  1  0]  [-1  0  1  0]  [ 1  0 -1  1]  [ 0  1  0  0]
[ 0  0  1  0]  [ 0  0  1  0]  [ 0  1 -1  1]  [ 0  1 -1  1]
[ 0  0  0  1], [ 0  0  0  1], [ 0  0  0  1], [ 0  0  0  1],

[ 1  0  0  0]  [ 1  0  0  0]  [-1  1  0  0]  [ 1  0  0  0]
[ 1 -1  1  0]  [ 0  1  0  0]  [ 0  1  0  0]  [ 1 -1  1  0]
[ 1 -1  0  1]  [ 0  1  0 -1]  [ 0  0  1  0]  [ 0  0  1  0]
[ 0  0  0  1], [ 0  0  1 -1], [ 0  0  1 -1], [ 0  0  1 -1],

[ 1  0  0  0]  [ 0 -1  1  0]  [ 0  0 -1  1]  [-1  1  0  0]
[ 0  1  0  0]  [-1  0  1  0]  [ 1  0 -1  1]  [-1  1 -1  1]
[ 0  1 -1  1]  [ 0  0  1  0]  [ 0  1 -1  1]  [ 0  1 -1  1]
[ 0  1 -1  0], [ 0  0  0  1], [ 0  0  0  1], [ 0  0  0  

We have that $A_4 \le GL_4(\mathbb{R})$ is represented here as 120 matrices - let's count the number of elements of each order:

**TECHNICAL NOTE** (i.e. note for myself to add to Sage later): here, the order of an element of a finite Coxeter group is not implemented, we lift to the ring of matrices

In [3]:
from collections import Counter
Counter( [ i.order() for i in MatrixGroup( [ i.matrix().change_ring(ZZ) for i in A4 ] ) ] )

Counter({4: 30, 2: 25, 5: 24, 3: 20, 6: 20, 1: 1})

We read this as: 
 - there are 30 elements of order 4
 - 25 elements of order 2
 - 24 elements of order 5
 - 20 elements of order 6
 - 1 element of order 1 (the identity)

Note that if we look at $S_5$, we have the same counts of elements with the same orders:

In [4]:
Counter([ i.order() for i in SymmetricGroup(5) ])

Counter({4: 30, 2: 25, 5: 24, 6: 20, 3: 20, 1: 1})

It turns out that we have that $A_n \cong S_{n+1}  $. We can show this via Sage's interface to [GAP](https://docs.gap-system.org/), a software package specifically for working with groups. Here, we utilize the powerful [StructureDescription](https://docs.gap-system.org/doc/ref/chap39.html#X8199B74B84446971) which attempts to describe the group given to it.

In [5]:
for n in (2..9):
    An = CoxeterGroup(["A", n], implementation = "permutation") # represent the Coxeter group internally as a permutation group
    print(f"A{n} ≅ {An.gap().StructureDescription()}")

A2 ≅ S3
A3 ≅ S4
A4 ≅ S5
A5 ≅ S6
A6 ≅ S7
A7 ≅ S8
A8 ≅ S9
A9 ≅ S10


We also have the presentation: $$S_n = A_{n-1} = \left\langle \; \substack{  s_i^2 \\ ( s_i s_{i+1}) ^3  \\ (s_i s_j)^2  \; |i-j|>1  } \; \right\rangle$$

### $B_n$
We attempt the same with the family named B:

In [6]:
for n in (2..5):
    Bn = CoxeterGroup(["B", n] , implementation = "permutation") # represent the Coxeter group internally as a permutation group
    print(f"|B{n}| = {len(Bn)}, B{n} ≅ {Bn.gap().StructureDescription()}")

|B2| = 8, B2 ≅ D8
|B3| = 48, B3 ≅ C2 x S4
|B4| = 384, B4 ≅ ((((C2 x C2 x C2) : (C2 x C2)) : C3) : C2) : C2
|B5| = 3840, B5 ≅ C2 x ((C2 x C2 x C2 x C2) : S5)


These groups are much more complicated - we have the following small examples.

**NOTE**: In what follows (following GAP's notation) let
 - $\mathbb{Z}_n$ be the cyclic group of order $n$ (represented in GAP as `Cn`)
 - $D_{2n}$ be the dihedral group with $|D_{2n}| = 2n$
 - $S_{n}$ be the symmetric group
 - $A \times B$ be the direct product of two groups $A$ and $B$  (in GAP, this is denoted `A x B`)
 - $N \rtimes H$ be the *semidirect* product of two groups where $N \trianglelefteq H$ is normal (in GAP, this is denoted `N : H`)


From the Coxeter graph, we have following more complicated presentation:

$$B_n = \left\langle \; \substack{  s_i^2 \\ \\ (s_1 s_2)^4 \\  ( s_i s_{i+1}) ^3 \; \; i \ne  1  \\ (s_i s_j)^2  \; |i-j|>1  } \; \right\rangle$$

For small cases, we have the following:
$$\begin{array}{|c|c|c|}
\hline
n=2 & |B_n| = 8 & B_n \cong D_8 \\
\hline
n=3 & |B_n| = 48 & B_n \cong \mathbb{Z}_2 \times S_4\\
\hline
n=4 & |B_n| = 384 & B_n \cong (((\mathbb{Z_2}^3 \rtimes \mathbb{Z}_2) \rtimes \mathbb{Z}_3) \rtimes \mathbb{Z}_2) \rtimes \mathbb{Z}_2 \\
\hline
n=5 & |B_n| = 3840 & B_n \cong \mathbb{Z}_2 \times \left( Z_2^4 \rtimes S_5 \right) \\
\hline
\end{array}$$
We can also look up the orders of these groups on the OEIS:

In [7]:
oeis( [ len( CoxeterGroup(["B", n]))  for n in [2..10] ] )

0: A000165: Double factorial of even numbers: (2n)!! = 2^n*n!.

### $D_n$
We now look at the infinite family $D$.

In [8]:
for n in (4..8):
    Dn = CoxeterGroup(["D", n] , implementation = "permutation") # represent the Coxeter group internally as a permutation group
    print(f"|D{n}| = {len(Dn)}, D{n} ≅ {Dn.gap().StructureDescription()}")

|D4| = 192, D4 ≅ (((C2 x C2 x C2) : (C2 x C2)) : C3) : C2
|D5| = 1920, D5 ≅ (C2 x C2 x C2 x C2) : S5
|D6| = 23040, D6 ≅ (C2 x C2 x C2 x C2 x C2) : S6
|D7| = 322560, D7 ≅ (C2 x C2 x C2 x C2 x C2 x C2) : S7
|D8| = 5160960, D8 ≅ (C2 x C2 x C2 x C2 x C2 x C2 x C2) : S8


The last entries above suggest we look at the orders of $D_n$ for larger $n$.

In [9]:
oeis( [ len( CoxeterGroup(["D", n]))  for n in [4..16] ] )

0: A002866: a(0) = 1; for n > 0, a(n) = 2^(n-1)*n!.
1: A032107: Number of reversible strings with n labeled beads of 2 colors.
2: A052594: E.g.f. x(1+x-2x^2)/(1-2x).

It seems that we have that $D_{n} \cong \mathbb{Z}_2^{n-1} \rtimes S_n$ 

**QUESTION**: What is the nature of this semidirect product? i.e. what is the $ \phi : \mathbb{Z}_2^{n-1} \to \text{Aut}(S_n)$ needed here?

It may be helpful to do this for the smallest case $D_4$ but I will leave this for another time...

### $I_2(m)$
These groups are a lot smaller than the ones treated above - we have that these are the dihedral groups (i.e. that  $ I_n \cong D_{2n} $ ) 

In [10]:
for n in (4..8):
    In = CoxeterGroup(["I", n], implementation="matrix" )
    # Represent these as matrix groups
    matrices = [ m.matrix() for m in list(In) ]
    gap_repr = MatrixGroup(matrices).gap().StructureDescription()
    print(f"|I{n}| = {len(In)}, I{n} ≅ {gap_repr}")

|I4| = 8, I4 ≅ D8
|I5| = 10, I5 ≅ D10
|I6| = 12, I6 ≅ D12
|I7| = 14, I7 ≅ D14
|I8| = 16, I8 ≅ D16


### Other Families

We now look at the Coxeter groups coming from the remaining irreducible Coxeter diagrams. We are given the following presentation:

$$ F_4 = \left\langle\ \{ r,s,t,u \}: \substack{ r^2, s^2, t^2, u^2 \\ (rs)^3, (st)^4, (tu)^3 \\ (rt)^2, (ru)^2, (su)^2 } \right\rangle  $$

Is this a finite group? We show this computationally - if it is, we should be able to calculate the group's order:


In [11]:
R.<r,s,t,u> = FreeGroup() # First construct the free group on r,s,t,u
relations = [
    r^2, s^2, t^2, u^2,
    (r*s)^3, (s*t)^4, (t*u)^3,
    (r*t)^2, (r*u)^2, (s*u)^2
]
len(R / relations) # Quotient R by the relations

1152

However, note that adding one more generator makes the group infinite; the following code times out:

```python
R.<r,s,t,u,v> = FreeGroup() # First construct the free group on r,s,t,u
relations = [
    r^2, s^2, t^2, u^2, v^2,
    (r*s)^3, (s*t)^4, (t*u)^3, (u*v)^3,
    (r*t)^2, (r*u)^2, (s*u)^2
]
Q = R / relations
Q.cardinality(limit=10**6)
```